### DISCUSSION TITLE
Queries in RPCs behaving differently (slowly) when using auth.uid()

### DISCUSSION THREAD

**Discussion by iddogino** [supabase/supabase#29518] [Questions]
I have the following set of views defined on top of the Stripe wrapper:

```sql
CREATE MATERIALIZED VIEW "public"."PricePlan" AS
    SELECT price.id, product.name
    FROM "stripe"."prices" AS price
    JOIN "stripe"."products" AS product ON product.id = price.product
    WITH DATA;

CREATE OR REPLACE VIEW "public"."Subscription" AS
    SELECT sub.*, product.name AS plan_name
    FROM "stripe"."subscriptions" AS sub 
    JOIN "public"."PricePlan" AS product ON product.id = sub.attrs->'plan'->>'id';
```

I have defined a function to query it for the current authenticated user like so:

```sql
CREATE OR REPLACE FUNCTION get_user_subscription() RETURNS setof "public"."Subscription" AS $$
    SELECT * FROM "public"."Subscription"
    WHERE customer = auth.uid()::text
$$ LANGUAGE SQL STABLE SECURITY DEFINER;
```

But it was EXTREMELY slow. When I went into the Stripe logs, I noticed it loads ALL the subscription (which required a lot of API calls due to pagination) rather than look up the subscription by the customer id. But if I hard coded a user id instead of using `auth.uid()::text`, it properly looked it up by the ID.

I then rewrote it to the following, which worked much faster by performing a Stripe lookup by the ID rather than loading all subscriptions:
```sql
CREATE OR REPLACE FUNCTION get_user_subscription() RETURNS setof "public"."Subscription" AS $$
declare user_id text := auth.uid()::text;
begin
    return query SELECT * FROM "public"."Subscription"
    WHERE customer = user_id; 
  end;
$$ LANGUAGE plpgsql SECURITY DEFINER;
```

Is this expected behavior / side effect of using `auth.uid()`, or a bug? This seems similar (but not exactly the same) as this issue: https://github.com/orgs/supabase/discussions/8733

**Comment by silentworks** [comment]
You should take a look at the performance improvement tips in this guide https://supabase.com/docs/guides/database/postgres/row-level-security#call-functions-with-select

**Comment by solvionsolutions** [comment]
Nice debugging on this one — the `declare user_id text := auth.uid()::text` trick works for exactly the reason you'd hope, and there's a lighter-weight version that keeps it inline. When you write `WHERE customer = auth.uid()::text`, Postgres treats `auth.uid()` as volatile and re-evaluates it per row, so it can't push the filter down — which with the Stripe wrapper means it pulls every subscription and paginates through the whole API instead of doing an ID lookup. Wrapping it in a scalar subquery gets you the same one-time evaluation your PL/pgSQL variable does, but without leaving SQL:

```sql
WHERE customer = (select auth.uid())::text
```

That hoists it into an initPlan that runs once per statement, and the wrapper can then do the direct lookup. It's the same pattern that matters on regular RLS policies too — on an indexed column Supabase measured a filter go from ~179ms to ~9ms just from the `(select ...)` wrap, so if `customer` isn't indexed on the underlying table, add that as well. Even though this is answered, leaving it here for anyone who lands on it from search.

I put together the full breakdown with the query plans here: https://www.guardlayer.io/blog/supabase-rls-performance
